pennylane/beta/qnode.py

```diff
@@ -0,0 +1,540 @@
+        provides_jacobian = device.capabilities().get("provides_jacobian", False)
+
+        if not provides_jacobian:
+            raise qml.QuantumFunctionError(
+                f"The {device.short_name} device does not provide a native "
+                "method for computing the jacobian."
+            )
+
+        return "device", {}, device
+
+    @staticmethod
+    def _validate_parameter_shift(device):
+        model = device.capabilities().get("model", None)
+
+        if model == "qubit":
+            return qml.gradients.param_shift, {}, device
+
+        if model == "cv":
+            return qml.gradients.param_shift_cv, {"dev": device}, device
+
+        raise qml.QuantumFunctionError(
+            f"Device {device.short_name} uses an unknown model ('{model}') "
+            "that does not support the parameter-shift rule."
+        )
+
+    @property
+    def tape(self):
+        """The quantum tape"""
+        return self._tape
+
+    qtape = tape  # for backwards compatibility
+
+    def construct(self, args, kwargs):
+        """Call the quantum function with a tape context, ensuring the operations get queued."""
+
+        if self.interface == "autograd":
+            # HOTFIX: to maintain compatibility with core, here we treat
+            # all inputs that do not explicitly specify `requires_grad=False`
+            # as trainable. This should be removed at some point, forcing users
+            # to specify `requires_grad=True` for trainable parameters.
+            args = [
+                qml.numpy.array(a, requires_grad=True) if not hasattr(a, "requires_grad") else a
+                for a in args
+            ]
+
+        self._tape = qml.tape.JacobianTape()
+
+        with self.tape:
+            self._qfunc_output = self.func(*args, **kwargs)
+
+        params = self.tape.get_parameters(trainable_only=False)
+        self.tape.trainable_params = qml.math.get_trainable_indices(params)
+
+        if not isinstance(self._qfunc_output, Sequence):
+            measurement_processes = (self._qfunc_output,)
+        else:
+            measurement_processes = self._qfunc_output
+
+        if not all(isinstance(m, qml.measure.MeasurementProcess) for m in measurement_processes):
+            raise qml.QuantumFunctionError(
+                "A quantum function must return either a single measurement, "
+                "or a nonempty sequence of measurements."
+            )
+
+        if not all(ret == m for ret, m in zip(measurement_processes, self.tape.measurements)):
+            raise qml.QuantumFunctionError(
+                "All measurements must be returned in the order they are measured."
+            )
+
+        for obj in self.tape.operations + self.tape.observables:
+
+            if getattr(obj, "num_wires", None) is qml.operation.WiresEnum.AllWires:
+                # check here only if enough wires
+                if len(obj.wires) != self.device.num_wires:

```

Comment:
Some operations might specify "I only apply to _all_ wires of a device" by setting the class attribute

```python
class SomeOp(Operation):
    num_wires = qml.operation.WiresEnum.AllWires
```

If this is the case, we check if this is indeed satisfied.

It would be more natural to have checks like this on the tape, not the QNode, since the tape is in charge of constructing and validating circuits. _However_, the issue is the tape doesn't have the device! Only the QNode has that.

Reply:
Oh yes, I missed that class attribute. Thanks for pointing it out!